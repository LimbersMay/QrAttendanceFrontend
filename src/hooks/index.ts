export * from './useForm';
export * from './useUiStore';
export * from './useAuthStore';
export * from './useCheckAuth';
export * from './useGroupStore';
export * from './useQrCodeStore';
export * from './useRegistryStore';
export * from './useQrAttendanceStore';