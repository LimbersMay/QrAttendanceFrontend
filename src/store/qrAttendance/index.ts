export * from "./group";
export * from "./qrCode";
export * from "./registry";
