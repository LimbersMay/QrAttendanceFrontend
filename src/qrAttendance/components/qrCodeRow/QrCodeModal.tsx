import {FormEvent} from "react";
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';
import {Divider, Grid, MenuItem, Select, SelectChangeEvent, TextField} from "@mui/material";
import {QrCodeDatePicker} from "./QrCodeDatePicker";
import {useUiSlice} from "../../../hooks/useUiSlice";
import {useQrCodeSlice} from "../../../hooks/useQrCodeSlice";
import {useForm} from "../../../hooks/useForm";
import {QrCode} from "../../interfaces";
import {ModalLayout} from "../ModalLayout";
import {useGroupSlice} from "../../../hooks/useGroupSlice";

const initialState: QrCode = {
    id: "",
    formId: "",
    url: "",
    groupId: "",
    name: "",
    date: "2023-02-18T02:19:01.000Z",
    enabled: false
}

export const QrCodeModal = () => {

    const {activeQrCode, handleUpdateQrCode, handleStartNewQrCode} = useQrCodeSlice();
    const { active: activeGroup } = useGroupSlice();
    const {toggleQrCodeModal, isQrCodeModalOpen} = useUiSlice();

    const {formState, onInputChange, onDateChange} = useForm(activeQrCode || initialState);
    const {name, date, enabled} = formState;

    const onSubmit = (event: FormEvent) => {
        event.preventDefault();

        if (!activeGroup) return;

        if (!activeQrCode) {
            handleStartNewQrCode(name, date, enabled);
            toggleQrCodeModal();
            return;
        }

        handleUpdateQrCode({
            ...activeQrCode,
            name,
            date,
            enabled
        });

        toggleQrCodeModal();
    }

    const handleSelectChange = ({target}: SelectChangeEvent) => {
        onDateChange(target.name, target.value === 'YES');
    }

    return (
        <ModalLayout condition={isQrCodeModalOpen} handleClose={toggleQrCodeModal}>
            <form onSubmit={onSubmit}>
                <Grid container>

                    <Grid item xs={12} sx={{textAlign: "center"}}>
                        <Typography variant="h6">Editing {activeQrCode?.name}</Typography>
                        <Divider/>
                    </Grid>

                    <Grid item xs={12} sx={{mt: 2}}>
                        <TextField
                            label="QR code name"
                            name="name"
                            value={name}
                            fullWidth
                            variant={"outlined"}
                            placeholder="name"
                            onChange={onInputChange}>
                        </TextField>
                    </Grid>

                    <Grid item xs={12} sx={{mt: 2}}>
                        <QrCodeDatePicker date={date} onChangeDate={(value: string) => onDateChange("date", value)}/>
                    </Grid>

                    <Grid item xs={12} sx={{mt: 2}}>
                        <Select name="enabled" fullWidth variant={'outlined'} value={enabled ? 'YES' : 'NO'}
                                onChange={handleSelectChange}>
                            <MenuItem value="YES">YES</MenuItem>
                            <MenuItem value="NO">NO</MenuItem>
                        </Select>
                    </Grid>

                    <Grid item xs={12} sx={{mt: 2}}>
                        <Button
                            variant="contained"
                            fullWidth
                            type={"submit"}
                        >
                            Save
                        </Button>
                    </Grid>
                </Grid>
            </form>
        </ModalLayout>
    );
}