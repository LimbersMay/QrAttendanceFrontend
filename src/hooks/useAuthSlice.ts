import {useAppDispatch, useAppSelector} from "../store";
import {checkingCredentials, login, logout, selectAuth} from "../store/auth";
import {qrAttendanceApi} from "../api/qrAttendanceApi";
import {onSetActiveGroup, setGroups, setQrCodes, setRegistries} from "../store/qrAttendance";

interface CreatingUserProps {
    name: string;
    email: string;
    lastname: string;
    password: string;
}

export const useAuthSlice = () => {

    const dispatch = useAppDispatch();
    const { displayName } = useAppSelector(selectAuth);

    const startCreatingUser = async ({name, email, lastname, password}: CreatingUserProps) => {
        // async code here
        dispatch(checkingCredentials());

        try {
            await qrAttendanceApi.post('/auth/register', {
                name,
                email,
                lastname,
                password
            });

            const response = await qrAttendanceApi.post('/auth/login', {
                email,
                password
            });

            const { body: user } = response.data;

            // sync code here
            dispatch(login({
                uid: user.id,
                displayName: user.name,
                email: user.email,
            }));
        } catch (error: any) {
            dispatch(logout(null));
        }
    }

    const startLogin = async (email: string, password: string) => {
        dispatch(checkingCredentials());

        try {
            const response = await qrAttendanceApi.post('/auth/login', {
                email,
                password
            }, {
                withCredentials: true,
                headers: {
                    'Content-Type': 'application/json'
                }
            });

            const { body: user } = response.data;


            dispatch(login({
                uid: user.id,
                displayName: user.name,
                email: user.email,
            }));

        } catch (error: any) {
            dispatch(logout(null));
        }
    }

    const startLogout = async () => {
        await qrAttendanceApi.post('/auth/logout', {}, {
            withCredentials: true,
            headers: {
                'Content-Type': 'application/json'
            }
        });

        dispatch(logout(null));
        dispatch(onSetActiveGroup(null));
        dispatch(setGroups([]));
        dispatch(setQrCodes([]));
        dispatch(setRegistries([]));
    }

    return {
        // properties
        displayName,

        // methods
        startCreatingUser,
        startLogin,
        startLogout
    }
}