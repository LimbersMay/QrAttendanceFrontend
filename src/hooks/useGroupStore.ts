import {useAppDispatch, useAppSelector} from "../store";
import {
    selectGroup,
    onSetActiveGroup,
    updateGroup,
    addEmptyGroup,
    setGroups
} from "../store/qrAttendance";
import {Group} from "../qrAttendance/interfaces";
import {qrAttendanceApi} from "../api/qrAttendanceApi";

export const useGroupStore = () => {

    const dispatch = useAppDispatch();
    const {groups, active} = useAppSelector(selectGroup);

    const startUpdateGroup = async (group: Group) => {
        // async code here
        const {id, name} = group;

        await qrAttendanceApi.put('/group/update', {
            id: id,
            updatedFields: {
                name
            }
        });

        // sync code here
        dispatch(updateGroup(group));
    }

    const startNewGroup = async () => {

        // async code here
        const response = await qrAttendanceApi.post('/group/create', {
            name: 'Default'
        });

        const {body: group} = response.data;

        const newGroup = {
            id: group.id,
            date: group.createdAt,
            name: group.name
        }

        // sync code here
        dispatch(addEmptyGroup(newGroup));
        dispatch(onSetActiveGroup(newGroup));
    }

    const startLoadingGroups = async () => {
        // async code here
        const response = await qrAttendanceApi.get(`/group/all`);
        const {body: groups} = response.data;

        // sync code here
        const mappedGroups: Group[] = groups.map((group: any) => ({
            id: group.id,
            name: group.name,
            date: group.createdAt
        }));

        dispatch(setGroups(mappedGroups));
    }

    const setActiveGroup = (group: Group) => {
        dispatch(onSetActiveGroup(group));
    }

    return {
        // properties
        groups,
        active,

        // methods
        startUpdateGroup,
        setActiveGroup,
        startNewGroup,
        startLoadingGroups
    }
}
