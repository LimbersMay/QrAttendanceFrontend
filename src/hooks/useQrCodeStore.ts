import {useAppDispatch, useAppSelector} from "../store";
import {
    selectQrCode,
    onSetActiveQrCode,
    updateQrCode, addEmptyQrCode, selectGroup, setQrCodes
} from "../store/qrAttendance";
import {QrCode} from "../qrAttendance/interfaces";
import {SnackbarUtilities} from "../utilities/snackbar-manager";
import {qrAttendanceApi} from "../api/qrAttendanceApi";
import {getEnvironments} from "../helpers/getEnvironments";

const { VITE_APIURL } = getEnvironments();

export const useQrCodeStore = () => {

    const dispatch = useAppDispatch();

    const { activeQrCode, qrCodes } = useAppSelector(selectQrCode);
    const { active } = useAppSelector(selectGroup);

    const setActiveQrCode = (qrCode: QrCode | null) => {
        dispatch(onSetActiveQrCode(qrCode));
    }

    const startUpdateQrCode = async (qrCode: QrCode) => {

        // async code here
        const {id, name, enabled, date} = qrCode;

        await qrAttendanceApi.put('/qrCode/update', {
            id: id,
            updatedFields: {
                name,
                enabled,
                manualRegistrationDate: date
            }
        });

        // sync code here
        dispatch(updateQrCode(qrCode));

        SnackbarUtilities.success(`QR Code ${qrCode.name} updated successfully`);
    }

    const startNewQrCode = async (name: string, manualRegistrationDate: string, enabled: boolean) => {

        const id = active?.id || '';

        const response = await qrAttendanceApi.post('/qrCode/create', {
            groupId: id,
            name,
            enabled,
            manualRegistrationDate,
            url: `${VITE_APIURL}/checkIn`
        });

        const {body} = response.data;

        const newQrCode: QrCode = {
            groupId: id,
            id: body.id,
            url: body.url,
            formId: body.formId,
            name: body.name,
            date: body.manualRegistrationDate,
            enabled: false
        }

        // sync code here
        dispatch(addEmptyQrCode(newQrCode));
    }

    const startLoadingQrCodes = async () => {
        // async code here
        const response = await qrAttendanceApi.get('/qrCode/all');
        const {body} = response.data;

        const qrCodes: QrCode[] = body.map((qrCode: any) => {
            return {
                id: qrCode.id,
                url: qrCode.url,
                formId: qrCode.formId,
                groupId: qrCode.groupId,
                name: qrCode.name,
                date: qrCode.manualRegistrationDate,
                enabled: qrCode.enabled
            }
        });

        // sync code here
        dispatch(setQrCodes(qrCodes));
    }

    return {
        // properties
        activeQrCode,
        qrCodes,

        // methods
        setActiveQrCode,
        startUpdateQrCode,
        startNewQrCode,
        startLoadingQrCodes
    }
}
