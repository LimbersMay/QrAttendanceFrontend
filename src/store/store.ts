import {AnyAction, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {authSlice} from "./auth";
import {groupSlice, qrCodeSlice, registrySlice} from "./qrAttendance";
import socketMiddleware from "./middlewares/socketMiddleware";
export const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        group: groupSlice.reducer,
        qrCode: qrCodeSlice.reducer,
        registry: registrySlice.reducer
    },
    middleware: [socketMiddleware]
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    AnyAction
>
