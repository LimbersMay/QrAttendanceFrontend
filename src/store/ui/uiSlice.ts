import {createSlice} from '@reduxjs/toolkit';
import {RootState} from "../store";

interface uiInitialState {
    isQrCodeModalOpen: boolean;
    isRegistryModalOpen: boolean;
    isShowingQrCode: boolean;
    isTitleModalOpen: boolean;
}

const initialState: uiInitialState = {
    isQrCodeModalOpen: false,
    isRegistryModalOpen: false,
    isShowingQrCode: false,
    isTitleModalOpen: false
}

export const uiSlice = createSlice({
    name: 'ui',
    initialState,
    reducers: {
        onShowQrCode: (state) => {
            state.isShowingQrCode = true;
        },
        onHideQrCode: (state) => {
            state.isShowingQrCode = false;
        },
        onToggleQrCodeModal: (state) => {
            state.isQrCodeModalOpen = !state.isQrCodeModalOpen;
        },
        onToggleRegistryModal: (state) => {
            state.isRegistryModalOpen = !state.isRegistryModalOpen;
        },
        onToggleTitleModal: (state) => {
            state.isTitleModalOpen = !state.isTitleModalOpen;
        }
    }
});


// Action creators are generated for each case reducer function
export const { onToggleQrCodeModal, onShowQrCode, onHideQrCode, onToggleRegistryModal, onToggleTitleModal } = uiSlice.actions;
export const selectUi = (state: RootState) => state.ui;