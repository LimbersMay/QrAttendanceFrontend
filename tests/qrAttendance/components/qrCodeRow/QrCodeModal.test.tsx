import {Provider} from "react-redux";
import {configureStore} from "@reduxjs/toolkit";
import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import {QrCodeModal} from "../../../../src/qrAttendance/components/qrCodeRow/QrCodeModal";
import {uiSlice} from "../../../../src/store/ui/uiSlice";
import {groupSlice, qrCodeSlice} from "../../../../src/store/qrAttendance";
import {withQrCodeModalOpenTrue} from "../../../fixtures/uiStates";
import {withActiveGroupAndGroupsState, withQrCodesState} from "../../../fixtures/qrAttendanceStates";
import * as useQrCodeStore from "../../../../src/hooks/useQrCodeStore";
import * as useUiStore from "../../../../src/hooks/useUiStore";
import * as useGroupStore from "../../../../src/hooks/useGroupStore";

const store = configureStore({
    reducer: {
        ui: uiSlice.reducer,
        qrCode: qrCodeSlice.reducer,
        group: groupSlice.reducer

    },
    preloadedState: {
        ui: {...withQrCodeModalOpenTrue},
        qrCode: {...withQrCodesState},
        group: {...withActiveGroupAndGroupsState}
    }
});

describe('tests for <QrCodeModal />', () => {

    const mockStartUpdateQrCode = jest.fn();
    const mockToggleQrCodeModal = jest.fn();
    const mockStartNewQrCode = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    test('should match snapshot', () => {

        const {container} = render(
            <Provider store={store}>
                <QrCodeModal/>
            </Provider>
        );

        expect(container).toMatchSnapshot();
    });

    test('should show the qr code name in the modal', () => {

        render(
            <Provider store={store}>
                <QrCodeModal/>
            </Provider>
        );

        const qrCodeName = screen.getByTestId('qrCode-title');

        expect(qrCodeName.textContent).toContain(`${withQrCodesState.activeQrCode?.name}`);
    });

    test('should call startUpdateQrCode and toggleQrCodeModal when submit the form if there is an active group', () => {

        const spyUseQrCodeStore = jest.spyOn(useQrCodeStore, 'useQrCodeStore');
        const spyUseUiStore = jest.spyOn(useUiStore, 'useUiStore');
        const spyUseGroupStore = jest.spyOn(useGroupStore, 'useGroupStore');

        (spyUseQrCodeStore as jest.Mock).mockReturnValue({
            activeQrCode: withQrCodesState.qrCodes[0],
            startUpdateQrCode: mockStartUpdateQrCode
        });

        (spyUseUiStore as jest.Mock).mockReturnValue({
            toggleQrCodeModal: mockToggleQrCodeModal,
            isQrCodeModalOpen: true
        });

        (spyUseGroupStore as jest.Mock).mockReturnValue({
            active: withActiveGroupAndGroupsState.active
        });

        render(
            <QrCodeModal/>
        );

        const modalForm = screen.getByTestId('qrCode-modal-form');
        fireEvent.submit(modalForm);

        waitFor(async () => {
            expect(mockToggleQrCodeModal).toHaveBeenCalled();
            expect(mockStartUpdateQrCode).toHaveBeenCalled();
        });
    });

    test('should call startNewQrCode and toggleQrCodeModal when submit the form if there is no active group', () => {

        const spyUseQrCodeStore = jest.spyOn(useQrCodeStore, 'useQrCodeStore');
        const spyUseUiStore = jest.spyOn(useUiStore, 'useUiStore');
        const spyUseGroupStore = jest.spyOn(useGroupStore, 'useGroupStore');

        (spyUseQrCodeStore as jest.Mock).mockReturnValue({
            activeQrCode: withQrCodesState.qrCodes[0],
            startNewQrCode: mockStartNewQrCode
        });

        (spyUseUiStore as jest.Mock).mockReturnValue({
            toggleQrCodeModal: mockToggleQrCodeModal,
            isQrCodeModalOpen: true
        });

        (spyUseGroupStore as jest.Mock).mockReturnValue({
            active: null
        });

        render(
            <QrCodeModal/>
        );

        const modalForm = screen.getByTestId('qrCode-modal-form');
        fireEvent.submit(modalForm);

        waitFor(async () => {
            expect(mockToggleQrCodeModal).toHaveBeenCalled();
            expect(mockStartNewQrCode).toHaveBeenCalled();
        });

    });
});
