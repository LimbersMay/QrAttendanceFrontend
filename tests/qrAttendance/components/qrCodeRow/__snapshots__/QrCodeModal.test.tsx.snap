// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <QrCodeModal /> should match snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
